CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift

```diff
@@ -232,15 +231,38 @@ public let DECL_NODES: [Node] = [
           "A `where` clause that places additional constraints on generic parameters like `where Element: Hashable`.",
         isOptional: true
       ),
+    ]
+  ),
+
+  Node(
+    kind: .actorDecl,
+    base: .decl,
+    nameForDiagnostics: "actor",
+    traits: [
+      "DeclGroup",
+      "NamedDecl",
+      "WithAttributes",
+      "WithGenericParameters",
+      "WithModifiers",
+    ],
+    children: [
+      Child(
+        name: "actorHeader",
```

Comment:
If there is anything in DeclGroupSyntax that allows access to the header, it needs to have a different name from the header children in all of the concrete nodes since it will need to have a broader type.

Reply:
I realized that as well as I continued reviewing and think this is an argument for not having a `DeclGroupSyntax` node.

Reply:
An alternative I've been considering is making the `DeclGroupHeader` node *non*-base and having it be something like:

```
  Node(
    kind: .declGroupHeader,
    base: .syntax,
    nameForDiagnostics: "declaration group header",
    parserFunction: "parseDeclarationGroupHeader",
    traits: [
      "WithAttributes",
      "WithModifiers",
      "DeclGroupHeaderProtocol", // slight naming issue I'll have to resolve
    ],
    children: [
      Child(
        name: "header",
        kind: .nodeChoices(choices: [
          Child(name: "actorHeader", kind: .node(kind: .actorDeclHeader)),
          Child(name: "classHeader", kind: .node(kind: .classDeclHeader)),
          Child(name: "enumHeader", kind: .node(kind: .enumDeclHeader)),
          Child(name: "extensionHeader", kind: .node(kind: .extensionDeclHeader)),
          Child(name: "protocolHeader", kind: .node(kind: .protocolDeclHeader)),
          Child(name: "structHeader", kind: .node(kind: .structDeclHeader)),
        ])
      )
    ]
  )
```

I would then add `modifiers`, `attributes`, etc. using a handwritten extension switching over the choices, so there'd be no existential overhead. Not quite as elegant, but it still gives us a concrete type that can represent any of the headers.

Reply:
Do we really need a `DeclGroupHeader` node at all? Correct me if I’m wrong, but I think the primary use for it is to be an option to `ABIAttributeArgumentsSyntax`, which already lists all sorts of other declarations. If we listed all the nodes that are subtypes of `DeclGroupHeaderSyntax` in there as well, I don’t think there’s real need for `DeclGroupHeaderSyntax` anymore. Or am I missing something?

Reply:
I also use it in the new implementation for `HasTrailingMemberDeclBlock`, and I believe I use it in ASTGen to generalize some of the lowering of DeclGroups, but those probably aren't mandatory—they're just nice little tricks I was able to do now that the new nodes existed.

Reply:
I think none of those are sufficient reason to introduce `DeclGroupHeaderSyntax`.